### PR TITLE
Clean up `FilterToolBar` logic

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

- Centralise logic from `FilterRevisionsHelper` and `FilterBranchHelper` in `FilterToolBar`
- Remove dependency in `FormBrowse` on `FilterRevisionsHelper` and `FilterBranchHelper`
- Remove dependency in `FormFileHistory` on `FilterRevisionsHelper` and `FilterBranchHelper`
- Deprecate and remove `FilterRevisionsHelper` and `FilterBranchHelper`

This will allow for a consistent UI experience in `FormBrowse` and `FormFileHistory`, as well as make it easy to enhance or modify the filter toolbar (e.g. add "show all branches", "show current branch", "show reflogs", etc. buttons).

Suggest to review commit by commit. Can be squashed in the end.

## Screenshots 

No UI changes

## Test methodology 

- Manual. Call me a hypocrite but right now I don't have any time to write tests for this...

### CODE REVIEW

**Review by gerhardol — COMMENTED** [review]
+1
Have played around briefly and briefly reviewed.
Note that there is one UI change, _Advanced filter_ button in FileHistory.
I do not really use the filters myself, so no deep usability testing - but there should be none.

No deep comments on the changes, they look good to me.

I do miss integration with the *Advanced filter* in `FormRevisionFilter`. The name on `RevisionFilter` is too close, maybe the form should change names as it is also handling branches and paths. (The path handling is only handled there.)
I would like to have some improvements there (some thoughts in #7598).
Maybe later.

Not sure if I will have more feedback, can approve if required to go on, but prefer if someone else look at it.

File: GitUI/UserControls/RevisionFilter.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿namespace GitUI.UserControls
+{
+    public class RevisionFilter
+    {
+        // TODO: consider using a bit vector or an enum
```
Comment (mstv) [inline GitUI/UserControls/RevisionFilter.cs:5]: or make it a `record`

**Review by mstv — COMMENTED** [review]
+1 after brief review

No more toolbar flicker. 🎉

One testcase failed:
https://ci.appveyor.com/project/gitextensions/gitextensions/builds/40276777/tests

File: GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs [hunk 2]
```diff
@@ -339,7 +339,7 @@ private ToolStripItem[] CreateCopyContextMenuItems()
 
         private void FilterInRevisionGrid(BaseBranchNode branch)
         {
-            _filterBranchHelper?.SetBranchFilter(branch.FullPath, refresh: true);
+            _branchFilterAction(branch.FullPath, true);
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs:342]: No can't do, it is an delegate
image

File: GitUI/UserControls/FilterToolBar.cs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            finally
+            {
+                _isApplyingFilter = false;
+            }
+        }
+
+        public void Bind(Func<IGitModule> getModule, IRevisionGridFilter revisionGridFilter)
+        {
+            _getModule = getModule ?? throw new ArgumentNullException(nameof(getModule));
+            _revisionGridFilter = revisionGridFilter ?? throw new ArgumentNullException(nameof(revisionGridFilter));
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/UserControls/FilterToolBar.cs:117]: This is a public API, so in theory it is possible that one day someone will call it with `null`... It is a small price to pay for maintainability sake.

File: GitUI/UserControls/RevisionFilter.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/UserControls/RevisionFilter.cs:5]: Yeah-nah :) I'm not planning to do it now.... There are other low hanging fruits we can pluck before this one.

**Comment by RussKie** [comment]
> Note that there is one UI change, _Advanced filter_ button in FileHistory.

I thought it was intended. I'm happy to remove it, though I think it is useful. I plan to add other filtering options (e.g. add "show all branches", "show current branch", "show reflogs", etc.) to the toolbar, which are useful on both forms.

> I do miss integration with the _Advanced filter_ in `FormRevisionFilter`.

Can you please elaborate?

> No more toolbar flicker. 🎉

That's an unexpected side effect :)

**Comment by gerhardol** [comment]
> > Note that there is one UI change, _Advanced filter_ button in FileHistory.
> 
> I thought it was intended. I'm happy to remove it, though I think it is useful. I plan to add other filtering options (e.g. add "show all branches", "show current branch", "show reflogs", etc.) to the toolbar, which are useful on both forms.

It is fine, this was in response to the PR description that states no UI changes.

> > I do miss integration with the _Advanced filter_ in `FormRevisionFilter`.
> 
> Can you please elaborate?

There are for instance author in both Advanced filter and the quick filter, so no good way to get a summary of the applied filters.
For branch this is direct duplication, the handling of other filters are a little more complicated.

**Comment by RussKie** [comment]
> > > I do miss integration with the _Advanced filter_ in `FormRevisionFilter`.
> > 
> > 
> > Can you please elaborate?
> 
> There are for instance author in both Advanced filter and the quick filter, so no good way to get a summary of the applied filters.
> For branch this is direct duplication, the handling of other filters are a little more complicated.

Let's handle this separately.

**Comment by RussKie** [comment]
I'd like to have it merged, if there are no objections.

**Comment by mstv** [comment]
> I'd like to have it merged, if there are no objections.

OK for me. Though #9454 should be merged first, shouldn't it?

**Comment by gerhardol** [comment]
> I'd like to have it merged, if there are no objections.

No objection from me

**Comment by RussKie** [comment]
> Though #9454 should be merged first, shouldn't it?

They are not really related, besides f93eb75

**Pull request merged**